onnxruntime/core/platform/windows/logging/etw_sink.cc

```diff
@@ -137,28 +137,47 @@ void NTAPI EtwRegistrationManager::ORT_TL_EtwEnableCallback(
 EtwRegistrationManager::~EtwRegistrationManager() {
   std::lock_guard<OrtMutex> lock(callbacks_mutex_);
   callbacks_.clear();
-  ::TraceLoggingUnregister(etw_provider_handle);
+  if (initialized_ || initializing_) {
+    std::lock_guard<OrtMutex> init_lock(init_mutex_);
+      assert(!initializing_);
+      if (initialized_) {
+        ::TraceLoggingUnregister(etw_provider_handle);
+        initialized_ = false;
+      }
+  }
 }
 
 EtwRegistrationManager::EtwRegistrationManager() {
 }
 
-void EtwRegistrationManager::LazyInitialize() {
-  if (!initialized_) {
+void EtwRegistrationManager::LazyInitialize() try {
+  if (!initialized_ && !initializing_) {
     std::lock_guard<OrtMutex> lock(init_mutex_);
-    if (!initialized_) {  // Double-check locking pattern
-      initialized_ = true;
+    if (!initialized_ && !initializing_) {  // Double-check locking pattern
+      initializing_ = true;
       etw_status_ = ::TraceLoggingRegisterEx(etw_provider_handle, ORT_TL_EtwEnableCallback, nullptr);
       if (FAILED(etw_status_)) {
         ORT_THROW("ETW registration failed. Logging will be broken: " + std::to_string(etw_status_));
       }
+      initialized_ = true;
+      initializing_ = false;
     }
   }
+} catch (...)
+{
+  initialized_ = false;
+  initializing_ = false;
```

Comment:
If `TraceLoggingRegisterEx` fails for any reason and throws, then it's unlikely to succeed the next time, because whatever bad condition caused trace logging to fail in the first place is likely persistent. So I worry this will repeatedly attempt to call `LazyInitialize` again and repeatedly throw, which would be worse (perfwise, given how expensive exceptions are) than just throwing once never attempting to initialize again. So If the state had four conditions (`NotInitialized`, `Initializing`, `Initialized`, `Failed`), and only attempted to initialize if it was `== NotInitialized`, that would be less noisy and clearer to read than two separate variables.